### PR TITLE
Feature/updated python3

### PR DESCRIPTION
**Pull request by prabhuramachandran**
This cherry-picks Pradyun's changes and fixes conflicts.  All tests pass on Python3.4 for me with this PR.

### CODE REVIEW

**Comment by sjagoe** [comment]
Please get Python 3.4 tests running on Travis CI

**Comment by sjagoe** [comment]
By the test eggs added, it looks like we need travis to test py2.6, py2.7 and py3.3+

File: envisage/developer/code_browser/code_browser.py [hunk 1]
```diff
@@ -2,15 +2,20 @@
 
 
 # Standard library imports.
-import cPickle, imp, logging, os, stat, warnings
+import sys, imp, logging, os, stat, warnings
+if sys.version_info[0] >= 3:
```
Comment (sjagoe) [inline envisage/developer/code_browser/code_browser.py]: This, along with similar checks are duplicated throughout the PR. Please create an `envisage._compat` module collecting this logic and import the required objects from there.

File: envisage/plugins/ipython_shell/view/namespace_view.py [hunk 2]
```diff
@@ -213,7 +213,7 @@ def _get_tree_editor(self):
             tree editor currently used.
         """
         for editor in self.ui._editors:
-            print editor
+            print(editor)
```
Comment (sjagoe) [inline envisage/plugins/ipython_shell/view/namespace_view.py]: Missing the `print_function` future import.

File: envisage/plugins/remote_editor/communication/tests/test_communication.py [hunk 3]
```diff
@@ -69,8 +69,9 @@ def testCommunication(self):
         self.assert_(os.path.exists(LOCK_PATH))
 
         # Test normal operation
-
-        self.assert_(Server.ping(get_server_port()))
+        tmp = get_server_port()
+        print(tmp)
```
Comment (sjagoe) [inline envisage/plugins/remote_editor/communication/tests/test_communication.py]: unintentional `print()`?

File: envisage/plugins/remote_editor/communication/util.py [hunk 4]
```diff
@@ -20,10 +24,25 @@
...
 
+# Python 3 compatibility
+PY3 = sys.version_info[0] >= 3
+STRING_BASE_CLASS = str if PY3 else basestring
+
+if PY3:
+    def encode(s):
+        return bytes(s, 'utf-8')
+    def decode(s):
+        return s.decode()
```
Comment (sjagoe) [inline envisage/plugins/remote_editor/communication/util.py]: Should this not be 'utf-8' as well?

Comment (sjagoe) [inline envisage/plugins/remote_editor/communication/util.py]: I suspect this will need a unicode-based test to ensure that it does actually work in both Python 2 and Python 3.

File: envisage/plugins/text_editor/editor/text_editor.py [hunk 5]
```diff
@@ -225,9 +225,9 @@ def _create_traits_ui_view(self):
     def _get_unique_id(self, prefix='Untitled '):
         """ Return a unique id for a new file. """
 
-        id = prefix + str(_id_generator.next())
+        id = prefix + next(str(_id_generator))
```
Comment (sjagoe) [inline envisage/plugins/text_editor/editor/text_editor.py]: `next` and `str` are in the wrong order.

File: envisage/plugins/text_editor/editor/text_editor.py [hunk 6]
```diff
@@ -225,9 +225,9 @@ def _create_traits_ui_view(self):
...
         while self.window.get_editor_by_id(id) is not None:
-            id = prefix + str(_id_generator.next())
+            id = prefix + next(str(_id_generator))
```
Comment (sjagoe) [inline envisage/plugins/text_editor/editor/text_editor.py]: `next` and `str` are in the wrong order.

File: envisage/provider_extension_registry.py [hunk 7]
```diff
@@ -103,10 +103,7 @@ def _get_extensions(self, extension_point_id):
         #     all = [x for y in extensions for x in y]
         #
         # But I'm sure sure that that makes it any clearer ;^)
-        all = []
-        map(all.extend, extensions)
-
-        return all
+        return [x for y in extensions for x in y]
```
Comment (sjagoe) [inline envisage/provider_extension_registry.py]: Please use better names than `x` and `y`, which will help with readability and understanding of the list comprehension.

File: envisage/resource/tests/resource_manager_test_case.py [hunk 8]
```diff
@@ -43,15 +51,15 @@ class ResourceManagerTestCase(unittest.TestCase):
     def setUp(self):
         """ Prepares the test fixture before each test method is called. """
 
-        self.stored_urlopen = urllib2.urlopen
-        urllib2.urlopen = stubout_urlopen
+        self.stored_urlopen = url_library.urlopen
```
Comment (sjagoe) [inline envisage/resource/tests/resource_manager_test_case.py]: Can this be converted to `mock`?

File: envisage/resource/tests/resource_manager_test_case.py [hunk 9]
```diff
@@ -74,7 +82,7 @@ def test_file_resource(self):
         f.close()
 
         # Open the api file via the file system.
-        g = file(filename, 'rb')
+        g = open(filename, 'rb')
```
Comment (sjagoe) [inline envisage/resource/tests/resource_manager_test_case.py]: Please with `with open() as g`.

File: envisage/tests/egg_basket_plugin_manager_test_case.py [hunk 10]
```diff
@@ -177,6 +179,7 @@ def on_broken_plugin(ep, exc):
         )
 
         ids = [plugin.id for plugin in plugin_manager]
+        print(ids)
```
Comment (sjagoe) [inline envisage/tests/egg_basket_plugin_manager_test_case.py]: Extra print.

File: envisage/tests/service_test_case.py [hunk 11]
```diff
@@ -54,6 +54,7 @@ class PluginB(Plugin):
         application = TestApplication(plugins=[a, b])
         application.start()
 
+
```
Comment (sjagoe) [inline envisage/tests/service_test_case.py]: Extra blank line.

File: envisage/ui/tasks/preferences_pane.py [hunk 12]
```diff
@@ -66,8 +66,8 @@ def trait_context ( self ):
     def apply(self, info=None):
         """ Handles the Apply button being clicked.
         """
-        trait_names = filter(self._model._is_preference_trait,
-                             self._model.trait_names())
+        trait_names = list(filter(self._model._is_preference_trait,
+                             self._model.trait_names()))
```
Comment (sjagoe) [inline envisage/ui/tasks/preferences_pane.py]: pep8

File: envisage/plugins/text_editor/editor/text_editor.py (hunk 6 above)
Comment (sjagoe) [inline envisage/plugins/text_editor/editor/text_editor.py]: I suspect this means bad test coverage in this area :-(

File: envisage/resource/tests/resource_manager_test_case.py (hunk 8 above)
Comment (prabhuramachandran) [inline envisage/resource/tests/resource_manager_test_case.py]: Will add a dependency, is that OK?  Could this be done in a separate PR?

Comment (sjagoe) [inline envisage/resource/tests/resource_manager_test_case.py]: You're right, this can be left alone for now; it is a separate issue. Probably not worth another immediate PR either.

**Comment by prabhuramachandran** [comment]
@sjagoe -- I think I have made all the changes from your review.  I added a simple test for the utf-8 encoding in the remote_editor communication package.  I have turned on Python 3.4 builds on travis-ci and all tests pass on all versions of Python.  The Python 3.4 test failures are not allowed.  Please let me know if there is anything else to be done.  Thanks!

**Comment by sjagoe** [comment]
@prabhuramachandran I'll take a look as soon as I am able; I have just seen your comment as I need to leave the office :-(

**Comment by prabhuramachandran** [comment]
@sjagoe -- no worries!

**Comment by prabhuramachandran** [comment]
bump. :)

**Comment by sjagoe** [comment]
Thanks for the reminder; I meant to look at this yesterday but it slipped my mind.

**Comment by prabhuramachandran** [comment]
No worries at all.  Sorry for pestering.
